### PR TITLE
SLT-800: Improve Cron logging

### PR DESCRIPTION
**Pull request by ragnarkurmwunder**
- Generate statistics at the end of the execution
- Generate stats as JSON
- Capture job output, and export also in the JSON

The logs will look like this:

```JSON
{
  "cron_name":"world:world-cli",
  "elapsed_time_s":0.02,
  "peak_mem_kb":7360,
  "exit_code":0,
  "output":"..."
}
```

### CODE REVIEW

File: charts/frontend/templates/services-cron.yaml [hunk 1]
```diff
@@ -43,13 +43,59 @@ spec:
...
-            command: ["/bin/sh", "-c"]
+            command: ["/bin/bash", "-c"]
             args:
               - |
-                 set -ex
-                 echo "starting cron run"
-                 {{ $job.command | nindent 18 }}
-                 echo "cron run completed"
+                # We need to save the executable command(s) into a file,
+                # because the 'time' can exeute only command-files,
```
Comment (badrange) [inline charts/frontend/templates/services-cron.yaml:50]: I think `exeute` should be `execute`

File: charts/frontend/templates/services-cron.yaml [hunk 2]
```diff
@@ -43,13 +43,59 @@ spec:
...
+                # $$ = pid of current running shell.
+                stats="/tmp/stats.$$"
+                output="/tmp/output.$$"
+                output_json="/tmp/output_json.$$"
+
+                # For backwards compatibility.
+                echo "starting cron run"
+
+                # The execution of the actual cron command.
+                # For backwards compatibility reasons sent the output to stdout as well (tee).
```
Comment (badrange) [inline charts/frontend/templates/services-cron.yaml:70]: `sent` -> `send`?

File: charts/frontend/templates/services-cron.yaml (hunk 1 above)
Comment (ragnarkurmwunder) [reply charts/frontend/templates/services-cron.yaml]: thanks, fixed.

File: charts/frontend/templates/services-cron.yaml (hunk 2 above)
Comment (ragnarkurmwunder) [reply charts/frontend/templates/services-cron.yaml]: thanks, fixed.

File: charts/frontend/templates/services-cron.yaml [hunk 3]
```diff
@@ -43,13 +43,59 @@ spec:
...
+
+                # For backward compatibility.
+                echo "cron run completed"
+
+                # It's tricky to get the exit code of a random command from a pipeline.
+                # It is highly shell-dependent.
+                exit_code="${PIPESTATUS[0]}"
+
+                # Convert the command output to JSON.
+                node -e 'const fs = require("fs"); console.log(JSON.stringify(fs.readFileSync(0, "utf-8")));' < "${output}" > "${output_json}"
```
Comment (Rade333) [inline charts/frontend/templates/services-cron.yaml:82]: Here I need to turn to @Jancis or @tharna with my question:

Can we _safely assume_ that `node` and `bash` (line 46) are always present in the services? Technically I guess it is possible to use the frontend chart but define a Dockerfile that does not use our silta-node base image.

**Review by Rade333 — COMMENTED** [review]
Need to get clarification for pending questions.

File: charts/frontend/templates/services-cron.yaml (hunk 3 above)
Comment (Jancis) [reply charts/frontend/templates/services-cron.yaml:82]: Correct, `node` might not be absent. There are services that run a different base image

**Review by Jancis — CHANGES_REQUESTED** [review]
This needs to be an opt-in feature, can't be enabled by default. See https://github.com/wunderio/drupal-project-k8s/pull/557#pullrequestreview-1188032718

**Comment by Rade333** [comment]
Closing due to inactivity.

**Pull request closed without merging**